### DISCUSSION TITLE
Implementing Attention Transfer for Object Detection

### DISCUSSION THREAD

**Discussion by ai21mtech13002** [yoshitomo-matsubara/torchdistill#256] [Q&A] [answered]
Hi Yoshitomo,

1. I am trying to implement Attention Transfer for Object Detection using FasterRCNN -Resnet50 as Teacher and FasterRCNN - Resnet18 as Student.

2. The Attention Transfer paper considers attention maps to be of matching spatial dimensions . I checked the output dimensions for the blocks of Resnet in each case.

Resnet - 50 Backbone (Output shape at Blocks 1 -4) (FORMAT - Batch Size, Channels, Width, Height)
<img width="288" alt="Screenshot 2022-11-28 at 1 06 58 PM" src="https://user-images.githubusercontent.com/104313548/204220189-281eac01-91ec-4524-bba1-bdb3d3a03edf.png">

Resnet -18 Backbone (Output shape at Blocks 1 -4) (FORMAT - Batch Size, Channels, Width, Height)

<img width="305" alt="Screenshot 2022-11-28 at 1 07 55 PM" src="https://user-images.githubusercontent.com/104313548/204220343-32cb3e10-258b-4b78-851e-b263c20496f9.png">

3. I assume the difference in the channels can be overcome by replicating each channel by 4 times. But ,
     (a) What should be the strategy to match the Width and Height so as to properly implement Attention Transfer.
     (b) Do we need to make any changes to the yaml config file for implementing the matching strategy or do we need to make changes to the student model architecture for the same?

Thanks.

**Comment by yoshitomo-matsubara** [comment, marked as answer]
Hi @ai21mtech13002 

> (a) What should be the strategy to match the Width and Height so as to properly implement Attention Transfer.

It's a sort of limitation of Attention Transfer.
In this case, you can introduce right after `layer1`, `layer2`, `layer3`, and `layer4` of ResNet-18, auxiliary convolution layer(s) whose outputs will not be processed by subsequent layers in the student model, but those auxiliary modules are used just to be "hooked" by forward hook to extract its output. 
However, it may be no longer the original AT, but more like a combination of AT + FitNet (1st stage) as FitNet also uses an auxiliary convolution layer for dimension matching if necessary (See [the blue module in Fig. 1 (b)](https://arxiv.org/pdf/1412.6550.pdf)).

> (b) Do we need to make any changes to the yaml config file for implementing the matching strategy or do we need to make changes to the student model architecture for the same?

If you want to introduce multiple auxiliary convolution layers to the student model, yes. You'll need to define a temporal wrapper class (which will be used only during training) like those in https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py
And then, you can define your added wrapper in the yaml file like [this](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/multi_stage/ft/resnet18_from_resnet34.yaml#L73-L87).

Hope this helps

**Reply by ai21mtech13002** [reply]
Thanks Yoshitomo,

Do you have any links that I can follow in order to create these auxillary conv layers for the Teacher (Resnet -50)?

**Reply by ai21mtech13002** [reply]
Also, can you clarify which methods do ce,ghnd and ktaad refer to? Their config files are present in the sample folder?

**Reply by yoshitomo-matsubara** [reply]
> Do you have any links that I can follow in order to create these auxillary conv layers for the Teacher (Resnet -50)?

A similar wrapper design can be found [here](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py#L291-L321).
This is a wrapper model for student model (or teacher model), where you first execute `forward` without changing the model interface (as if you just run your model as is), and then post_forward function will be executed so that you can extract intermediate outputs (say layer1, ... layer4) from `io_dict` and feed the extracted output to auxiliary modules you'll define in the wrapper class.
By registering hooks in a yaml file for outputs from the auxiliary modules executed in `post_forward`, you can use the outputs to match those from teacher model.


> Also, can you clarify which methods do ce,ghnd and ktaad refer to? Their config files are present in the sample folder?

CE: Cross-Entropy (all the sample task folders except ImageNet)
GHND: Generalized Head Network Distillation (ImageNet and COCO 2017)
KTAAD: Knowledge Transfer and Adaptation (See [this sample config](https://github.com/yoshitomo-matsubara/torchdistill/tree/main/configs/sample/coco2017/multi_stage/ktaad))

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

1. I have added auxillary conv layers to the output of each layer (1 -4 ) in Resnet -18 model to match the outputs of Resnet-50 layers 1-4. The modified Resnet -18 model is attatched.
[fasterrcnn_R18_FPN.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10121308/fasterrcnn_R18_FPN.txt)


3. I have also suitably modified the yaml config file like this. 
[config.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10121306/config.txt)

4. However I am getting the following error if I keep a batch size of 1
<img width="1125" alt="Screenshot 2022-11-30 at 4 16 08 PM" src="https://user-images.githubusercontent.com/104313548/204776267-fc2241d6-af10-4242-a739-81bc36a443d3.png">

and the following if I keep batchsize greater than 1:

<img width="1425" alt="Screenshot 2022-11-30 at 2 20 03 PM" src="https://user-images.githubusercontent.com/104313548/204777402-15fa3e9a-f02f-4706-9fe9-0c15a32e32ac.png">



Can you guide where I may be going wrong.

P.S. : I don't think it should be a problem with the dataset, since I was using the same dataset while implementing Factor Transfer.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 
At first glance, the problem is that the layers you added to the model are not auxiliary, the output of the new layers will be propagated to the subsequent layers and change its behavior.

At high-level, you need to do something like this
```
...
z1 = self.layer1(z)
aux_z1 = self.aux_layer1(z1)
z2 = self.layer2(z1)
aux_z2 = self.aux_layer2(z2)
...
```

And the wrapper scheme I suggested above can achieve this without modifying your student model, by using the wrapper.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo, I went through the wrapper scheme. I have a few queries:

(a) What would be the input_module_path and how to specify it?
(b) What is the role of supp_dict and will it be required in my case?
(c) I understand that the io_dict is the same as specified in the demo python notebook on extracting intermediate representations. Am i right?
(d) Lastly, after creating the wrapper class, how do i include it in my model.py file?

Thanks.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

> (a) What would be the input_module_path and how to specify it?

`input_module_path` is str, a module path of the student model where you want to hook the input and/or output. You specify the path in the yaml file like [this](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/coco2017/multi_stage/ft/custom_fasterrcnn_resnet18_fpn_from_fasterrcnn_resnet50_fpn.yaml#L155). Note that the student model in this case is wrapped and the original student model is referred to as `student_model`.

> (b) What is the role of supp_dict and will it be required in my case?

supp_dict is used when you have other types of input to models like cached outputs from teachers. In your case, you don't need to worry about that.

> (c) I understand that the io_dict is the same as specified in the demo python notebook on extracting intermediate representations. Am i right?

Yes, it plays the same role

> (d) Lastly, after creating the wrapper class, how do i include it in my model.py file?

You just add [this python decorator](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/torchdistill/models/special.py#L111) right before your wrapper class and load the module in your main class so that you can specify your wrapper class in your yaml file

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo, I have created a wrapper class as you guided in special.py. Could you have a look at it and tell me if the implementation is correct? (I am planning to use a list containing the module paths to all the 4 layers of Student Resnet 18 and pass it as 'input_module_paths').

I am enclosing the wrapper class below.
[new_module.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10610268/new_module.txt)

Also, I am confused how to add forward hooks in the yaml file to the auxillary convolution layers added in the post forward method.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

You should move the following lines to `__init__`. 
```python
        #Defining the auxillary convolutions to match the shape of output of the teachers Resnet layers
        self.aux_conv_1=nn.Conv2d(self.layer1_out.shape[1],self.layer1_out.shape[1]*4,3,1,(1,21),bias=False)
        self.aux_conv_2=nn.Conv2d(self.layer2_out.shape[1],self.layer2_out.shape[1]*4,3,1,(1,11),bias=False)
        self.aux_conv_3=nn.Conv2d(self.layer3_out.shape[1],self.layer3_out.shape[1]*4,3,1,(1,6),bias=False)
        self.aux_conv_4=nn.Conv2d(self.layer4_out.shape[1],self.layer4_out.shape[1],(3,4),1,(1,4),bias=False)
```
Otherwise, `self.aux_conv_?` will be initialized every single time you feed a batch to the model.
In other words, you need to specify parameters for nn.Conv2d (in_channel and out_channel, you should know the parameters beforehand) to define the auxiliary convolution layers beforehand.

On top of that, you will want to wrap your `self.aux_conv_?` with DistributedDataParallel in case you use distributed training mode.
i.e., I would write `__init__` like this
```python
    def __init__(self, student_model, input_module_paths, aux_conv1_kwargs, aux_conv2_kwargs, aux_conv3_kwargs, aux_conv4_kwargs, device, device_ids, distributed, **kwargs):
        #Propose to pass a list of module paths 
        super().__init__()
        self.student_model = wrap_if_distributed(student_model, device, device_ids, distributed)
        self.input_module_paths = input_module_paths
        self.aux_conv_1 = wrap_if_distributed(nn.Conv2d(**aux_conv1_kwargs), device, device_ids, distributed)
        self.aux_conv_2 = wrap_if_distributed(nn.Conv2d(**aux_conv2_kwargs), device, device_ids, distributed)
        self.aux_conv_3 = wrap_if_distributed(nn.Conv2d(**aux_conv3_kwargs), device, device_ids, distributed)
        self.aux_conv_4 = wrap_if_distributed(nn.Conv2d(**aux_conv4_kwargs), device, device_ids, distributed)
```

To add forward hooks in the yaml file to the auxiliary convolution layers, you will write
```yaml
forward_hook:
  input: []
  output: ['aux_conv_1', 'aux_conv_2', 'aux_conv_3', 'aux_conv_4']
```

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo. I have made the desired changes in the wrapper class and also configured the yaml file. I am however getting the following error:-
<img width="1405" alt="Screenshot 2023-02-05 at 11 44 36 AM" src="https://user-images.githubusercontent.com/104313548/216804762-8985c1b1-1021-4629-9ff3-261bd755c5a9.png">

In the logs it mentions "No special module called `Student4AttentionTransfer` is registered."

I am also enclosing my yaml file 
[yaml.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10610328/yaml.txt)

and
the special.py file
[special.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10610330/special.txt)

Thanks and Regards

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

I guess that you directly edited the existing `torchdistill/models/special.py` and attempted to run `examples/object_detection.py`
If so, I'd suggest you refer to [this answer](https://github.com/yoshitomo-matsubara/torchdistill/discussions/278#discussioncomment-4700064) (and use `register_special_module` instead of `register_model_func`)

If not, please clarify how you installed torchdistill, and what files you added/edited.

Also, you will need to add `student_model.` as a prefix of the student model's module paths because with respect to the `Student4AttentionTransfer` wrapper class, the module paths of the original student model should start with "student_model." as you defined so `self.student_model = wrap_if_distributed(student_model, device, device_ids, distributed)`

i.e.,
```yaml
    special:
      type: 'Student4AttentionTransfer'
      params: 
        input_module_paths: ['student_model.backbone.body.layer1','student_model.backbone.body.layer2','student_model.backbone.body.layer3','student_model.backbone.body.layer4']
```

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo. I believe that the registry issue has been resolved. I created a new .py file and included the wrapper class there. 

I have also made necessary changes in the yaml file as under:-
[yaml.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10610392/yaml.txt)

Enclosing the separate .py file created under the examples folder.
[special.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10610380/special.txt)

 But I am getting new errors now as under:-
<img width="1426" alt="Screenshot 2023-02-05 at 12 20 28 PM" src="https://user-images.githubusercontent.com/104313548/216805900-687c5588-e514-4923-ad84-90a09c71e27f.png">

Thanks and Regards.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo, I was able to resolve the error by adding the following in the train/val and test dataloader configs in the yaml file.

collate_fn: 'coco_collate_fn'

However this is the latest error that i am getting.
<img width="1410" alt="Screenshot 2023-02-05 at 12 38 49 PM" src="https://user-images.githubusercontent.com/104313548/216806428-c913f883-9a3c-4a3f-b828-fbd091212d72.png">


Thanks and Regards.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

The error looks like occurring in `collate` process in DataLoader. Since your yaml file does not specify any `collate` function, it should be using PyTorch default collate function.

For COCO datasets, my object detection config files use `collate_fn: 'coco_collate_fn'`.
If your dataset is compatible with COCO format, you can try `collate_fn: 'coco_collate_fn'` for `*_data_loader` entries like [this](https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/coco2017/single_stage/ghnd/custom_maskrcnn_resnet50_fpn_from_maskrcnn_resnet50_fpn.yaml#L62)

Unfortunately, I am not familiar with the dataset you're using thus cannot give you any more concrete advice for this issue at this moment.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo. Yes , the collate issue is now resolved by adding collate_fn: 'coco_collate_fn'. 

There does not seem to be a problem with the dataset as I ran the factor transfer experiment with the same dataset and the same configs for dataset in the yaml file.

I believe the error is coming from the wrapper class added  which I had named at_module.py (as one can see the same in the logs). Do you have any insight why I would be getting this error.
<img width="1410" alt="Screenshot 2023-02-05 at 12 38 49 PM" src="https://user-images.githubusercontent.com/104313548/216806686-f99ae5d0-2662-412d-8bd8-d50cec059be4.png">

Thanks and regards.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

At first please identify where the error came from. Without much context of your code (repo), I cannot fully understand how it's supposed to work. Also, please use text for log and code instead of screenshot for better search so that others can find this thread if they have similar topics.

I believe your `self.student_model` does not get `targets` as input during training, and the errors are from your model `fasterrcnn_R18`.

Add [`forward_proc: 'forward_batch_target'`]( https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/coco2017/multi_stage/ft/custom_fasterrcnn_resnet18_fpn_from_fasterrcnn_resnet50_fpn.yaml#L133) to student model (and teacher model if necessary) when you need to pass `targets` to their forward function

**Reply by ai21mtech13002** [reply]
Thanks Yoshitomo,

The model is getting the targets now. However, there seems to be some issue with the AT loss calculation i believe as I am getting the following error.
[error_log.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10610595/error_log.txt)

Could you please check if the manner in which I have defined the pairs for AT loss is correct. Enclosing the yaml file.
[yaml.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10610596/yaml.txt)

I am using the similar additional python file to specify the student wrapper
[special.txt](https://github.com/yoshitomo-matsubara/torchdistill/files/10610597/special.txt)

Regards and Thanks.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo. I went ahead and tried to access the io_dict that was generating the error. So , it turns out that the io_dict for the auxillary convolution layer (aux_conv_1) defined in the separate student wrapper is returning an empty dict.
<img width="589" alt="Screenshot 2023-02-05 at 7 37 56 PM" src="https://user-images.githubusercontent.com/104313548/216824331-d7786d9f-a5dd-431c-a619-7232f249b3fc.png">

Does this mean that there is some error in the way we have defined the wrapper for the student? That is torchdistill is not able to acces the module_paths of the auxillary convolution layers?

Thanks and Regards.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

> Does this mean that there is some error in the way we have defined the wrapper for the student? That is torchdistill is not able to acces the module_paths of the auxillary convolution layers?

Again, please review your code and config before you have a doubt on torchdistill.

The yaml file misses forward hooks for layers you defined as `input_module_paths`.
You should add `'student_model.backbone.body.layer1','student_model.backbone.body.layer2','student_model.backbone.body.layer3','student_model.backbone.body.layer4'` to `output` in `forward` to register forward hooks and extract outputs of the layers.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

The experiment has run successfully. These were the final results.
<img width="1405" alt="Screenshot 2023-02-06 at 12 40 53 AM" src="https://user-images.githubusercontent.com/104313548/216839808-f3db2298-7128-401b-b4b3-750fa4ee673f.png">

With Attention Transfer the Student model achieves almost comparable performance with the Teacher Model. However, I found Factor Transfer to give better results.

Thanks for all the guidance and help.

Regards.

**Reply by ai21mtech13002** [reply]
Hi Yoshitomo,

Is there any sample config using which I can implement Knowledge Review using Torchdistill?

Regards.

**Reply by yoshitomo-matsubara** [reply]
Hi @ai21mtech13002 

It's here https://github.com/yoshitomo-matsubara/torchdistill/blob/main/configs/sample/ilsvrc2012/single_stage/kr/resnet18_from_resnet34.yaml

If you continue, please open a separate discussion as the topics in this tread are getting diverse.

**Reply by ai21mtech13002** [reply]
Thanks Yoshitomo.
